SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipLiveness.swift

```diff
@@ -0,0 +1,713 @@
+// verified. Remove uses of FunctionPassContext.
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+/// Compute liveness and return a range, which the caller must deinitialize.
+///
+/// `definingValue` must introduce an OSSA lifetime. It may be either
+/// an owned value or introduce a borrowed value (BeginBorrowValue),
+/// including:
+///
+/// 1. Owned non-phi values
+/// 2. Owned phi values
+/// 3. Borrow scope introducers: begin_borrow/load_borrow
+/// 4. Reborrows: guaranteed phi that ends its operand's borrow scope and
+///    requires post-dominating scope-ending uses (end_borrow or reborrow)
+///
+/// `definingValue`'s lifetime must already complete on all paths
+/// (a.k.a linear). Only lifetime-ending operations generate liveness.
+///
+/// `definingValue` dominates the range. Forwarding and phi uses do
+/// not extend the lifetime.
+///
+/// This is the simplest OSSA liveness analysis. It assumes that OSSA
+/// lifetime completion has already run on `definingValue`, and it
+/// cannot fix OSSA lifetimes after a transformation.
+func computeLinearLiveness(for definingValue: Value, _ context: Context)
+  -> InstructionRange {
+
+  assert(definingValue.ownership == .owned
+    || BeginBorrowValue(definingValue) != nil,
+    "value must define an OSSA lifetime")
+
+  // InstructionRange cannot directly represent the beginning of the block
+  // so we fake it with getRepresentativeInstruction().
+  let defInst =
+    InstructionRange.getRepresentativeInstruction(for: definingValue)
+  var range = InstructionRange(begin: defInst, context)
+
+  // Compute liveness.
+  definingValue.lifetimeEndingUses.forEach { range.insert($0.instruction) }
+
+  return range
+}
+
+/// Indicate whether OwnershipUseVisitor is visiting a use of the
+/// outer OSSA lifetime or within an inner borrow scope (reborrow).
+enum IsInnerLifetime {
+case outerLifetime
+case innerLifetime
+}
+
+typealias InnerScopeHandler = (Value) -> WalkResult
+
+/// Compute liveness and return a range, which rhe caller must deinitialize.
+///
+/// An OSSA lifetime begins with a single "defining" value, which
+/// must be owned, or must begin a borrow scope.
+///
+/// - Liveness does not extend beyond lifetime-ending operations
+/// (a.k.a. affine lifetimes).
+///
+/// - The definition dominates all use points.
+///
+/// - Does not assume the current lifetime is complete, but does
+/// assume any inner scopes are complete. Use `innerScopeHandler` to
+/// complete them or bail-out.
+func computeInteriorLiveness(for definingValue: Value,
+  _ context: FunctionPassContext,
+  _ innerScopeHandler: InnerScopeHandler? = nil) -> InstructionRange {
+
+  assert(definingValue.ownership == .owned
+    || BeginBorrowValue(definingValue) != nil,
+    "value must define an OSSA lifetime")
+
+  // InstructionRange cannot directly represent the beginning of the block
+  // so we fake it with getRepresentativeInstruction().
```

Comment:
This comment is gone now because we have `InstructionRange.init(for: Value)`. But the point is still valid. Any client that needs to reason about where the initial value is defined, at the entry of a block vs. the first instruction needs to maintain that information on the side along with the range.